### ISSUE TITLE
Feature request: Merging scale parameters (gamma) into weight matrix

### ISSUE THREAD

**Issue by NilsGraf** [jiachenzhu/DyT#15]
Here is a suggestion for DyT:  To simplify and further speed-up inference, would it make sense to merge the scale params (gamma) of DyT into the next weight matrix?

This is similar to the [FlashNorm arXiv paper](https://arxiv.org/abs/2407.09577) and [FlashNorm github repo](https://github.com/OpenMachine-ai/transformer-tricks)

![here](https://raw.githubusercontent.com/OpenMachine-ai/transformer-tricks/refs/heads/main/doc/fig/flashNorm_fig1.svg)

**Comment by jiachenzhu** [comment]
Hey, Nils,

I failed to understand how (b) is equivalent to (c)? Would you mind show me in math formula?

**Comment by NilsGraf** [comment]
Sure, keep in mind that we assume that the linear layer has no bias (as mentioned in the Flash Normalization paper). Also, matrix multiplication by a scalar is commutative.

Example for 2 x 2 weight matrix W with elements w11, w12, w21, w22 and scalar s:
y1 = x1 s w11 + x2 s w21
y2 = x1 s w12 + x2 s w22
Using commutative property, you can factor out s:
y1 = ( x1 w11 + x2 w21 ) s
y2 = ( x1 w12 + x2 w22 ) s
So you can perform the scaling by scalar s after doing the vector-matrix multiplication. If you have a bias, do the scaling by scalar s before adding the bias vector.

**Comment by NilsGraf** [comment]
Hi Jiachen, I forgot to mention the following: 

In your DyT paper, there is also a bias "beta" right after the scaling by "gamma", see equation (2) of your paper. If the DyT layer is followed by a linear layer with weight-matrix `W`, you can first eliminate your bias "beta" or `b` as follows, where vector `y` is the activation after scaling by gamma, i.e. `y = gamma ∗ tanh(alpha * x)` from your equation (2), and `c` is the bias of the linear layer, and `z` is the linear layer output.
```
z = (y + b) * W + c
  = y * W  + b * W + c
  = y * W  + (b * W + c) 
  = y * W + c'
```
Note that `b * W` is constant for inference, so it can be pre-computed off-line. So you can merge the normalization bias `b` into the bias `c` of the linear layer to get the new constant bias `c' = bW + c` at the output of the linear layer. 

So you can update your parameter file to eliminate `b` and use the modified `c'` instead, which simplifies inference and reduces the number of tensors in your safetensor file, similar to flash normalization. I'll update the flash normalization paper to mention this.

**Comment by jiachenzhu** [comment]
Hey @NilsGraf,

Thanks for the explanation — I now understand what you're doing. You're merging the learnable weights from the normalization layers with the subsequent linear weights. I believe torch.compile can already handle this case by fusing the weights automatically, but it is still a good suggestion for speedup the inference when torch.compile is not available.

Best,
Jiachen Zhu

**Comment by NilsGraf** [comment]
Hi @jiachenzhu,

Thanks for your note. I'm not sure that torch.compile is automatically fusing the norm-weights with linear layer weights: See for example [this article](https://www.adamcasson.com/posts/torch-compile-vit) which says that torch.compile currently doesn't perform parameter-fusion for a similar transformer model, below is copied from the conclusion section of this article:
>  ... Even though torch.compile has become a crucial component to running efficient PyTorch models, there can still be room for optimizations that it isn't quite able to find for itself [4](https://www.adamcasson.com/posts/torch-compile-vit#user-content-fn-he_tweet). In the case of this ViT architecture, one example is that the LayerScale weights (gamma) can be multiplied into the weights and biases of the linear layer before it in order to completely eliminate the multiplication op of LayerScale while maintaining mathematical equivalance.


As a side note:  It seems that torch.compile can perform parameter-fusion for Conv layers followed by BatchNorm for inference,  taking advantage of "frozen" mean and variance estimates during inference (which is not the case for layerNorm and RMSNorm for transformers), see the following links for more details:
- https://github.com/pytorch/pytorch/issues/108863
- https://nenadmarkus.com/p/fusing-batchnorm-and-conv/
- paper https://arxiv.org/abs/2305.11624
